### DISCUSSION TITLE
About Scripts constantly running on the background

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#544] [General]
(I'm creating this thread to split the discussion of #520. It was getting a multi-topic messy thread over there)

Next weekend we are going to work on the script documentation and I think this theme could be included, at some point. I think it's a relevant theme, because having a script doing actions on the background opens a lot of possibilities that can be of the interest of all kinds of users.  

I know this kind of script exists for a long time, but I don't see a systematization of the methods to achieve it, pointing the pros and cons of each approach.

I found those 3 ways of creating such scripts:

1. Node listeners
2. Script Filter inside the Map (and Node) conditional styles
3. [This "Multithread" thing](https://sourceforge.net/p/freeplane/discussion/758437/thread/7c6605c656/), which I don't understand, but I tested the script, and it works

What are the pros and cons of each one?
We discussed bit about it on #520. To me, it is still not clear, for example, why macmarrun said that "The proper approach would be to create a NodeChangeListener with a script" [instead of using a Script Filter?].

I think this thread can also be used as a place to concentrate the research about how those things work (like was being done at #520 )

**Comment by macmarrum** [comment]
@eu2021 Further to your comment [#520 discussioncomment-3341166](https://github.com/freeplane/freeplane/discussions/520#discussioncomment-3341166)

> Now, what about using the new getListeners() like this on a Script Filter:
> 
>     * IF getListeners() is not empty: return false
> 
>     * ELSE:  execute the scripts that have the listeners. And also create a dummy listener, just for the purpose of making the first condition true, which allows deactivating the other listeners if the user wants. return: true

I took your idea and created a helper library that executes only once per mindmap, so that it can go into a Node Conditional Style (Script Filter)

1. Put [ScriptOnMapOpen.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/ScriptOnMapOpen.groovy) in `<user.dir>/lib` and restart Freeplane
2. In your mindmap, add a Node Conditional Style to the root node, with Script Filter `ScriptOnMapOpen.nodeConditionalStyle(node)`
3. On the root node, add an attribute `scriptOnMapOpen` with the code to be executed when the mindmap is opened
4. Save the mindmap, close it and open again to enjoy the script being executed automagically

The accompanying library [MacmarrumChangeListenerUtils.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/MacmarrumChangeListenerUtils.groovy) contains a NodeChangeListener and a MapChangeListener, which can be initiated from `scriptOnMapOpen`, with the code `MacmarrumChangeListenerUtils.toggleChangeListeners(node)`

This example shows how and when a NodeChangeListener can be used and the difference to a MapChangeListener.

A Map/Node Conditional Style could be used too, but as mentioned previously, the functionality hasn't been meant for write access. In other words, Freeplane doesn't expect that a Script Filter will change anything in a mindmap; FP operates on the assumption that a Script Filter only reads the needed properties. On the other hand, a ChangeListener is designed for code to change its mindmap and gives the code's author control over what events should trigger the execution.

> I found those 3 ways of creating such scripts:
>    1. Node listeners
>    2. Script Filter inside the Map (and Node) conditional styles
>    3. [This "Multithread" thing](https://sourceforge.net/p/freeplane/discussion/758437/thread/7c6605c656/), which I don't understand, but I tested the script, and it works
> 
> 
> What are the pros and cons of each one? 

I would divide the points into two categories

#### Ways to automatically launch code in Freeplane on start or on map open
* init script – for the entire application (on Freeplane start)
* Script Filter – for an individual mindmap (on map open)

#### Ways to run code in the background
* Script Filter – on map change / on node change
  The code is usually executed many times for each node, because most of events in a map/node can affect the style of a node. The assumption is that the code only reads information, so it should be fairly quick and unobstructive.
* Map/Node Listener – on map change / on node change
  The code's author has the control over which events (onNodeInserted, onNodeMoved, onNodeDeleted, onNodeChanged:TEXT, etc) should trigger the code. This usually means less operations as compared to Script Filter. The author has the ability to ignore events that are triggered as a result of write operations done by the code itself (see `MacmarrumChangeListenerUtils.MacmarrumNodeChangeListener#canReact`)
* Thread ("multithread thing") – at a time interval (e.g. every 30 seconds)
  This approach ignores change events, and instead focuses on passing time. In other words, the elapse of time is the event. The code executes even if nothing (else) happens in a mindmap.

**Reply by euu2021** [reply]
Please, I'm failing at following your step by step instructions:

> 2. In your mindmap, add a Node Conditional Style to the root node, with Script Filter ScriptOnMapOpen.nodeConditionalStyle(node)

Does the content of the filter should be simply `ScriptOnMapOpen.nodeConditionalStyle(node)` , like this?
![image](https://user-images.githubusercontent.com/77707706/183537007-059e23b7-2c8e-49c0-a110-e71a8f7ed2e3.png)

If so, this is giving me this error (on a new with only the root):
![image](https://user-images.githubusercontent.com/77707706/183538359-37540dbc-6e5d-49ed-af74-6d0e552c0736.png)


> 3. On the root node, add an attribute scriptOnMapOpen with the code to be executed when the mindmap is opened
The name of the attribute should be `scriptOnMapOpen`? If so, I'm having a hard time pasting a code into the attribute. If simply paste on the lateral panel, it gets without line breaks:
![image](https://user-images.githubusercontent.com/77707706/183537771-4cee3e20-eef5-4214-a443-14adddc58e0b.png)


I tried inserting a simple one liner code, but it doesn't work, anyway. Maybe, the main problem is on step 2.

**Reply by macmarrum** [reply]
> Does the content of the filter should be simply ScriptOnMapOpen.nodeConditionalStyle(node) , like this?

Yes. You got it :+1: 

> If so, this is giving me this error (on a new with only the root):

I can see the error talks about a "failed compilation". It looks like the file you saved to `ScriptOnMapOpen.groovy` is incorrect. Please open `<user.dir>/lib/ScriptOnMapOpen.groovy` in an editor (notepad.exe is fine). I am guessing it contains the html page. It should contain only the Groovy code.
On https://github.com/macmarrum/freeplane-scripting/blob/main/src/ScriptOnMapOpen.groovy you can click the **Raw** button to display the raw file. It might be easier this way to save it to your disk.

Once you have this sorted out, in Freeplane you can use `Tools->Edit script…` to access `scriptOnMapOpen` attribute. NB any attribute that starts with `script` is visible in `Tools->Edit script…`.

Anyway, here's a sample mindmap that shows `scriptOnMapOpen` in action (provided `ScriptOnMapOpen.groovy` is installed).
[scriptOnMapOpen.zip](https://github.com/freeplane/freeplane/files/9288773/scriptOnMapOpen.zip)

**Reply by euu2021** [reply]
---
> I am guessing it contains the html page.

Yes, that was the problem. I always have a hard time downloading github files 🤦‍♂️.

>NB any attribute that starts with script is visible in Tools->Edit script…

Interesting. I didn't know this.

---

Ok, thanks. I got the examples working. 

Now, what about having multiple scripts creating listeners? I tried pilling up two scripts in the root, but only the first one gets executed (if I change the order, the other one gets executed):

![image](https://user-images.githubusercontent.com/77707706/183669782-a00d3dae-c181-4917-8c53-7c2b176e049c.png)

Isn't it possible? Is it necessary to have everything inside the same script?

---
So, just to make it clear what is the current situation: 

You propose a solution to the problem: how to make a script constantly run on the background, with an autostart that also covers the situations when a map is opened inside FP (ie, without restarting FP).

Your proposed solution requires all those 3 things: 

- a library inside `<user.dir>/lib`
- a Script filter on a node conditional style
- a script in a node, that creates a listener

**Comment by macmarrum** [comment]
> Ok, thanks. I got the examples working.

Cool :sunglasses:

> Now, what about having multiple scripts creating listeners? I tried pilling up two scripts in the root, but only the first one gets executed

The code responsible for getting the script from root's attribute is in line 28 of [ScriptOnMapOpen.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/ScriptOnMapOpen.groovy)
```groovy
def script = root['scriptOnMapOpen']?.text
```
Here's at the source code of `org.freeplane.plugin.script.proxy.NodeProxy#getAt`
```java
public Convertible getAt(final String attributeName) {
    final Object value = getAttributes().getFirst(attributeName);
    return ProxyUtils.attributeValueToConvertible(getDelegate(), getScriptContext(), value);
}
```
I can see that only the first attribute is returned, which makes sense.

But `ScriptOnMapOpen.groovy` can be changed to go over each attribute named `scriptOnMapOpen` and execute them one by one. Maybe this could be a good exercise for you to try?

Anyway, I think it is more convenient to write a Change Listener in an IDE and make it available as a class (library), like I did with [MacmarrumChangeListenerUtils.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/MacmarrumChangeListenerUtils.groovy). It can contain as many invocations of listeners as you want.

**Reply by euu2021** [reply]
> But ScriptOnMapOpen.groovy can be changed to go over each attribute named scriptOnMapOpen and execute them one by one. Maybe this could be a good exercise for you to try?

Nice. I will try.

I will open another reply about the actual making of the scripts.

**Comment by euu2021** [comment]
Another front on this topic is: creating an easy guide on how to make the node listener scripts.

I will try to understand your examples and create an easy template. I think a template like this will be helpful, pointing where the "usual script" must be inserted:

```
Code that doesn't need to be touched.

Kinds of changes to be watched <-- change here as you wish

if (change happens)
	do this <--- insert your usual script here
```
	
I will try to apply into #548 and see what I get.

> Anyway, I think it is more convenient to write a Change Listener in an IDE and make it available as a class (library), like I did with [MacmarrumChangeListenerUtils.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/MacmarrumChangeListenerUtils.groovy). It can contain as many invocations of listeners as you want.

Yes, IDE and classes are two things that I need to learn how to use in order to go to the next level of scripting.

**Comment by euu2021** [comment]
I found a 4th way of creating this kind of script:

4. Script Filter inside the Find toolbar, with `Filter->Highlight all matching nodes` activated. 

One big difference is that it works for any maps opened. For example, here it automatically changes "aaa" for "this" on every map I open (dangerous!):

![javaw_9HXtdxu6W9](https://user-images.githubusercontent.com/77707706/184151734-3800603d-1e05-4ec5-a70a-3dd9ae9d749c.gif)

It's kinda crazy, though, because the script keeps running even if the criteria in the filter toolbar is changed.

---
I also remembered this 5th way:

5. Keep Freeplane on a loop, as used on [this script](https://docs.freeplane.org/#/scripting/Scripts_collection?id=monitor-clipboard)
